[addons] add first step of gui on addon interface

AlwinEsch opened the pull request:
<!--- Provide a general summary of your change in the Title above -->

## Description
<!--- Describe your change in detail -->

This add the first parts of gui related addon callback parts to new
style. There becomes standard functions and the dialogs added.

On next request comes the controls list item and window.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

ksooo commented on xbmc/addons/interfaces/GUI/DialogKeyboard.cpp:
```diff
@@ -0,0 +1,216 @@
+  return CGUIKeyboardFactory::SendTextToActiveKeyboard(aTextString, closeKeyboard);
+}
+
+bool Interface_GUIDialogKeyboard::isKeyboardActivated(void* kodiBase)
```
For consistency reasons, the function name should start with an upper case, like all others do.

AlwinEsch replied:
comes, still working on the gui parts, there are few parts who bring me grey hairs, during improvement ;-)

ksooo commented on xbmc/addons/interfaces/GUI/DialogOK.cpp:
```diff
@@ -0,0 +1,74 @@
+  if (!addon)
+  {
+    CLog::Log(LOGERROR, "kodi::gui::DialogOK:%s - invalid data (addon='%p', heading='%p', text='%p')",
+                                        __FUNCTION__, addon, heading, text);
```
Indentation?!

ksooo commented on xbmc/addons/interfaces/GUI/DialogOK.cpp:
```diff
@@ -0,0 +1,74 @@
+  if (!addon)
+  {
+    CLog::Log(LOGERROR, "kodi::gui::DialogOK::%s - invalid data (addon='%p', heading='%p', line0='%p', line1='%p', line2='%p')",
+                                        __FUNCTION__, addon, heading, line0, line1, line2);
```
Indentation?!

ksooo commented on xbmc/addons/interfaces/GUI/General.cpp:
```diff
@@ -0,0 +1,730 @@
+    case ACTION_ANALOG_MOVE_X          : return ADDON_ACTION_ANALOG_MOVE_X;
+    case ACTION_ANALOG_MOVE_Y          : return ADDON_ACTION_ANALOG_MOVE_Y;
+
+    case ACTION_ERROR                  : return ADDON_ACTION_ERROR;
```
Could you please reformat the code to have the case on one line and thne return on the next line - here and in the rest of this source file? I think this is part of Kodi coding guidelines.

AlwinEsch replied:
In know but with action id's becomes it really long, but I can change it

ksooo replied:
Thanks.

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/DialogTextViewer.h:
```diff
@@ -0,0 +1,118 @@
+    ///  "but there were uncredited contributions by others. The songs were written "
+    ///  "by Edgar \"Yip\" Harburg (lyrics) and Harold Arlen (music). The incidental "
+    ///  "music, based largely on the songs, was composed by Herbert Stothart, with "
+    ///  "interspersed renderings from classical composers.\n");
```
Could you please shorten this example string? I guess the usage of this function gets clear without this wall of text. ;-)

AlwinEsch replied:
Was so big to see on example the full functionallity of TextViewer with scroll up and down.

And you know the Wizard of Oz is a big story :-D

ksooo replied:
Okay, you decide on this. :-)

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/definitions.h:
```diff
@@ -0,0 +1,1373 @@
+  int (*get_screen_width)(void* kodiBase);
+  int (*get_video_resolution)(void* kodiBase);
+  int (*get_current_window_dialog_id)(void* kodiBase);
+  int (*get_current_window_id)(void* kodiBase);
```
Why do the gui_general function names start with a lower case letter where all other gui function names start with an uppercase letter? This should be consistent.

AlwinEsch replied:
Is changed, was gone down during move from kodi agile to here.

ksooo reviewed (changes requested):
Epic. Some suggestions and questions, though.

AlwinEsch commented on xbmc/addons/interfaces/GUI/DialogFileBrowser.cpp:
```diff
@@ -0,0 +1,334 @@
+    for (unsigned int i = 0; i < entries; ++i)
+    {
+      if (path[i])
+        free(path[i]);
```
Have tried it on other requests and has several times a crash then. Don't know why.

AlwinEsch commented on xbmc/addons/interfaces/GUI/DialogFileBrowser.cpp:
```diff
@@ -0,0 +1,334 @@
+  free(addonInterface->toKodi->kodi_gui->dialogFileBrowser);
+}
+
+bool Interface_GUIDialogFileBrowser::ShowAndGetDirectory(void* kodiBase, const char *shares, const char *heading, char &path, unsigned int &maxStringSize, bool writeOnly)
```
Are changed.

AlwinEsch commented:
Is everything updated, buhh :D

Excuse me for the circumstances, with so many things I do not see the forest in front of loud trees.

AlwinEsch commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/definitions.h:
```diff
@@ -0,0 +1,1383 @@
+  /// Move left off a control.
+  ///
+  ADDON_ACTION_MOVE_LEFT               = 1,
+
```
Have someone a good idea about them for action id's?

To do with a separate place bring advantages but also disadvantages.

I remove them now and hope to find a good way to interact with them and kodi and hold also independent.

With the "enum" has made the doxygen output a bit nicer as with a "define". :-)

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/definitions.h:
```diff
@@ -0,0 +1,164 @@
+  bool (*show_and_get_source)(void* kodiBase, char* path, unsigned int maxStringSize, bool allowNetworkShares, const char* additionalShare, const char* type);
+  bool (*show_and_get_image)(void* kodiBase, const char* shares, const char* heading, char* path, unsigned int maxStringSize);
+  bool (*show_and_get_image_list)(void* kodiBase,  const char* shares, const char* heading, char*** path, unsigned int* entries);
+  void (*clear_list)(void* kodiBase, char*** path, unsigned int entries);
```
This function should be renamed to clear_file_list for consistency reasons.

ksooo commented on xbmc/addons/interfaces/GUI/DialogFileBrowser.cpp:
```diff
@@ -0,0 +1,338 @@
+    for (unsigned int i = 0; i < entries; ++i)
+    {
+      if ((*path)[i])
+        free((*path)[i]);
```
The <code>if ((*path)[i])</code> can be removed. <code>free</code> can be called safely on a nullptr.

AlwinEsch replied:
I test it

ksooo replied:
No need to test, specification is clear on this. ;-)

ksooo commented on xbmc/addons/interfaces/GUI/DialogFileBrowser.cpp:
```diff
@@ -0,0 +1,338 @@
+
+  if (*path)
+  {
+    for (unsigned int i = 0; i < entries; ++i)
```
Question: Is parameter <code>entries</code> actually needed or can the count of elements be calculated via <code>sizeof</code> here?

AlwinEsch replied:
Is needed, the allocated size of them can be different.

ksooo commented on xbmc/addons/interfaces/GUI/DialogFileBrowser.cpp:
```diff
@@ -0,0 +1,338 @@
+    CLog::Log(LOGERROR, "Interface_GUIDialogFileBrowser::%s - invalid handler data (path='%p') on addon '%s'", __FUNCTION__, path, addon->ID().c_str());
+}
+
+void Interface_GUIDialogFileBrowser::GetVECShares(VECSOURCES& vecShares, std::string strShares, std::string strPath)
```
<code>const std::string& strShares</code>, <code>const std::string& strPath</code> ?! No reason for pass-by-value, imo.

ksooo commented on xbmc/addons/interfaces/GUI/DialogFileBrowser.cpp:
```diff
@@ -0,0 +1,338 @@
+      if ((*path)[i])
+        free((*path)[i]);
+    }
+    free(*path);
```
<code>*path = 0;</code> is missing here. Otherwise the caller might operate on a pointer pointing to the woods.

ksooo reviewed (changes requested):
Still needs some more love. ;-)

AlwinEsch commented on xbmc/addons/interfaces/GUI/DialogFileBrowser.cpp:
```diff
@@ -0,0 +1,338 @@
+  GetVECShares(vecShares, shares, strPath);
+  bool bRet = CGUIDialogFileBrowser::ShowAndGetDirectory(vecShares, heading, strPath, writeOnly);
+  if (bRet)
+    strncpy(path, strPath.c_str(), maxStringSize);
```
I look what possible, problem on them is that the they become a `[in|out]` as values.
If it use malloc must be in from out different.

ksooo replied:
yeah, easy. Use two params, one for "in" (const char * pathIn), one for "out" (char *** pathOut).

AlwinEsch commented:
Is updated :)

ksooo commented on xbmc/addons/interfaces/GUI/DialogFileBrowser.cpp:
```diff
@@ -0,0 +1,341 @@
+}
+
+bool Interface_GUIDialogFileBrowser::show_and_get_file_list(void* kodiBase, const char* shares, const char* mask,
+                                                            const char* heading, char*** path, unsigned int* entries,
```
I would name the variable file_list, not path as it contains a list of several files, not a single path.

ksooo commented on xbmc/addons/interfaces/GUI/DialogFileBrowser.cpp:
```diff
@@ -0,0 +1,341 @@
+  return bRet;
+}
+
+bool Interface_GUIDialogFileBrowser::show_and_get_image_list(void* kodiBase, const char* shares, const char* heading, char*** path, unsigned int* entries)
```
I would name the variable file_list, not path as it contains a list of several files, not a single path.

ksooo commented on xbmc/addons/interfaces/GUI/DialogFileBrowser.cpp:
```diff
@@ -0,0 +1,341 @@
+  return bRet;
+}
+
+void Interface_GUIDialogFileBrowser::clear_file_list(void* kodiBase, char*** path, unsigned int entries)
```
I would name the variable file_list, not path as it contains a list of several files, not a single path.

ksooo reviewed (commented):
Last nitpicking, I promise. :-)

xhaggi commented on xbmc/addons/interfaces/GUI/DialogContextMenu.cpp:
```diff
@@ -0,0 +1,72 @@
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (!addon)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIDialogContextMenu::Open - invalid kodi base data");
```
Should we use `__FUNCTION__` as well here?

xhaggi commented on xbmc/addons/interfaces/GUI/DialogContextMenu.cpp:
```diff
@@ -0,0 +1,72 @@
+  CGUIDialogContextMenu* dialog = g_windowManager.GetWindow<CGUIDialogContextMenu>(WINDOW_DIALOG_CONTEXT_MENU);
+  if (!heading || !entries || !dialog)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIDialogContextMenu::Open - invalid handler data (heading='%p', entries='%p', dialog='%p') on addon '%s'",
```
same here

AlwinEsch committed:
[addons] add first step of gui on addon interface

This add the first parts of gui related addon callback parts to new
style. There becomes standard functions and the dialogs added.

On next request comes the controls list item and window.

ksooo commented:
jenkins build this please
